pkg/cache/cache.go

```diff
@@ -0,0 +1,97 @@
+package cache
+
+import (
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"go.uber.org/zap"
+)
+
+type Cache struct {
+	cacheDir string
+}
+
+func New(rootDir string) (*Cache, error) {
+	cacheDir := filepath.Join(rootDir, "cache")
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("creating a cache directory: %w", err)
+	}
+
+	return &Cache{cacheDir: cacheDir}, nil
+}
+
+func (cache *Cache) Get(fileIdentifier string) (path string, err error) {
+	path, err = cache.identifierPath(fileIdentifier)
+	if err != nil {
+		return "", fmt.Errorf("searching for identifier '%s' in cache: %w", fileIdentifier, err)
+	}
+
+	if !exists(path) {
+		return "", fs.ErrNotExist
+	}
+
+	return path, nil
+}
+
+func (cache *Cache) Put(fileIdentifier string, reader io.Reader) error {
+	path, err := cache.identifierPath(fileIdentifier)
+	if err != nil {
+		return fmt.Errorf("searching for identifier '%s' in cache: %w", fileIdentifier, err)
+	}
+
+	if exists(path) {
+		zap.S().Warnf("File with identifier '%s' already exists in cache", fileIdentifier)
+		return fs.ErrExist
+	}
+
+	zap.S().Infof("Storing file with identifier '%s' in cache", fileIdentifier)
+
+	file, err := os.Create(path)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
+	}
+	defer file.Close()
+
+	if _, err = io.Copy(file, reader); err != nil {
+		return fmt.Errorf("storing file: %w", err)
+	}
+
+	return nil
+}
+
+func (cache *Cache) identifierPath(fileIdentifier string) (string, error) {
+	identifier, err := identifierHash(fileIdentifier)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(cache.cacheDir, identifier), nil
+}
+
+func identifierHash(identifier string) (string, error) {
+	h := fnv.New64()
+
+	if _, err := h.Write([]byte(identifier)); err != nil {
+		return "", err
+	}
+
+	return strconv.FormatUint(h.Sum64(), 10), nil
+}
+
+func exists(path string) bool {
```

Comment:
I know you said this is a basic implementation, so I am not surprised to see the cache purely based on filename instead of more details like size or checksum. That said, it might not hurt to drop a comment here acknowledging that this is intentionally limited over time. That way, new developers don't look at this and wonder why we didn't flesh this out further.

Reply:
Isn't this more of a validation thing which should live elsewhere? Reason being is that not every file we cache will have a checksum to validate against.